iothub_client/tests/common_dt_e2e/iothubclient_common_dt_e2e.c

```diff
@@ -10,568 +10,728 @@
+    const unsigned char* value =  (unsigned char*)json_object_dotget_string(root_object,
+                                                                            full_property_name);
+    size_t length = json_object_dotget_string_len(root_object, full_property_name);
+    char* return_value = (char*)_malloc_and_copy_unsigned_char(value, length);
+
+    json_value_free(root_value);
+
+    return return_value;
+}
+
+static int _parse_json_twin_int(const char* twin_payload,
+                                const char* full_property_name,
+                                bool allow_for_zero)
+{
+
+    JSON_Value* root_value = json_parse_string(twin_payload);
+    ASSERT_IS_NOT_NULL(root_value);
+    JSON_Object* root_object = json_value_get_object(root_value);
+    ASSERT_IS_NOT_NULL(root_object);
+
+    double double_value = json_object_dotget_number(root_object, full_property_name);
+    int int_value = (int)(double_value + 0.1); // Account for possible underflow by small increment.
+
+    if (!allow_for_zero)
     {
-        result = IoTHubModuleClient_SetModuleTwinCallback(iothub_moduleclient_handle, twinCallback, device);
+        ASSERT_ARE_NOT_EQUAL(int, 0, int_value, "Failed to parse `%s`.", full_property_name);
     }
-    else
+
+    json_value_free(root_value);
+
+    return int_value;
+}
+
+static char* _parse_json_twin_char_from_array(const char* twin_payload,
+                                              const char* full_property_name,
+                                              size_t index)
+{
+    JSON_Value* root_value = json_parse_string(twin_payload);
+    ASSERT_IS_NOT_NULL(root_value);
+    JSON_Object* root_object = json_value_get_object(root_value);
+    ASSERT_IS_NOT_NULL(root_object);
+
+    JSON_Array* array = json_object_dotget_array(root_object, full_property_name);
+    ASSERT_IS_NOT_NULL(array, "Array not specified.");
+
+    const unsigned char* value =  (unsigned char*)json_array_get_string(array, index);
+    size_t length = json_array_get_string_len(array, index);
+    char* return_value = (char*)_malloc_and_copy_unsigned_char(value, length);
+
+    json_value_free(root_value);
+
+    return return_value;
+}
+
+static int _parse_json_twin_int_from_array(const char* twin_payload,
+                                           const char* full_property_name,
+                                           size_t index,
+                                           bool allow_for_zero)
+{
+    JSON_Value* root_value = json_parse_string(twin_payload);
+    ASSERT_IS_NOT_NULL(root_value);
+    JSON_Object* root_object = json_value_get_object(root_value);
+    ASSERT_IS_NOT_NULL(root_object);
+
+    JSON_Array* array = json_object_dotget_array(root_object, full_property_name);
+    ASSERT_IS_NOT_NULL(array, "Array not specified.");
+
+    double double_value = json_array_get_number(array, index);
+    int int_value = (int)(double_value + 0.1); // Account for possible underflow by small increment.
```

Comment:
I don't think we should cast to `int` - the standard mechanism is do the check as a double and then allow for a tolerance of rounding error.  The test code [here](https://github.com/jspaith/azure-sdk-for-c/blob/84bfe57b50091d88dc993ce80e95ca435aaa0d16/sdk/tests/iot/hub/test_az_iot_hub_client_properties.c#L1237) in embedded SDK demonstrates.